---
title: "Preparation for Session 1"
author: "Python Group"
format: 
    html:
        output-file: "session1.html"
        toc: true
        css: ../styles.css
        code-copy: true
        code-line-numbers: false
        link-external-icon: false
        link-external-newwindow: true
editor_options:
  markdown:
    wrap: sentence
    canonical: true
---

::: callout-tip
## **Learning objectives of the article**

Here is a breakdown of the key takeaways of the article:

**üí°Aim 1**: know key features of Python and its applications.

**üí°Aim 2**: know the steps to install Python through Miniconda/Anaconda Distribution.

**üí°Aim 3**: know what Python virtual environments are and how to use Conda to install and manage them.

**üí°Aim 4**: know features of the Visual Studio Code integrated development environments (IDE) and steps to set up a Python project within it.
:::

This is a required reading prior to the **Introduction to Python** workshop.
In this article, you can find things you need to know to get started with Python programming: ***How do I install Python? What is the best way to do it (if there is any)? What are other useful software tools to help make coding experience better?*** This article will answer these questions and serve as a follow-along guide for installing the necessary programs to help you prepare for the upcoming sessions.

On the other hand, this handout will touch on some important questions‚Äì***What is special about Python? And why do we want to learn it at all as biostatisticians?*** To answer these, we will first talk about the main features of Python, what it can do, and its applications in scientific research.
Later in the workshop, we will dive deeper into these topics and learn specific Python functions and packages/libraries through hands-on programming practices.

## üìñQuick overview of the workshop:

**Session 1** will help you set up a flexible, interactive, and reproducible environment for Python programming, including setting up **conda** virtual environments and Visual Studio Code IDE to write code in Jupyter Notebook

**Session 2 and 3** will focus on the Python language basics such as data structure, list comprehensive, and functions.
We will also learn the `pandas` library for data frame manipulation.

**Session 4** will introduce you to object-oriented programming (OOP)--the foundation of the Python language--and learn through example class objects from the statistics package `statsmodels` and machine learning library `scikit-learn`.

## ‚ùîWhat do I need for the workshop?

-   **Bring your laptop (Windows/MacOS)**

-   **Download the software**‚Äì*You* *will follow this article to install most of the software needed before coming to session 1.*

    ::: callout-important
    We recommend that you install software and run through the tutorial on your MSK laptop or workstation.
    Downloading/installing software files on VDI is extremely slow and might not install at all if the software is too large.
    :::

-   **Basic GitHub knowledge & MSK GitHub Enterprise account**‚ÄìThe workshop website is hosted on MSK Enterprise GitHub, which might require you logging in with MSK credentials to access links/download files.
    Check out [Biostatistics Resource Guide](https://rconnect.mskcc.org/resource-guide/101-trainings.html) for GitHub-related training.

# 1 Introduction to Python

## What is Python?

![](images/paste-1.png){width="220"}

> Python is a **high-level**, **interpreted**, and **general-purpose programming language** first developed by Guido van Rossum in 1991.

Python has gained much popularity in the past 20 years.
Its user group has expanded into a large and active scientific computing and developer community that spans numerous academic and industrial fields.
Nowadays, Python has a powerful ecosystem of [external packages (libraries)](https://pypi.org/) for data science, artificial intelligence, and software development.

Python is cross-platform and open-source.
In Python, you can easily install packages with the built-in installer `pip` or package manager `conda`, just as you do with `install.packages()` in R.

#### What can Python do?

Just like R, Python is an open source and versatile programming language that allows users to perform a wide range of data analysis and computational tasks.
While R is particularly useful in statistical analysis and visualizations, Python has been used in many distinct areas, such as:

-   Machine Learning and Deep learning

-   Web Development

-   Scripting & Automation

-   Cloud Computing

-   Game Development

-   Cybersecurity

## Why learn Python‚Äìas Biostatisticians?

Within the field of biostatistics/ bioinformatics, Python has become a core tool for biomedical data analysis due to its versatility, reproducibility, and strong ecosystem of scientific libraries.
Below are some areas where Python can be useful and some essential libraries.

#### **Statistical analysis**

While R is the go-to tool for ststistical analysis, Pyton has caught up with many equivalent libraries and functions:

-   `statsmodels`/ `scipy.stats` provide **regression modeling** and **hypothesis testing**.

-   `lifelines`/ `scikit-survival` support **survival analysis** and plotting.

#### **ML/DL ecosystem**

Python dominates in machine learning and AI development:

-   `scikit-learn` is a rich **machine learning** library that supports both supervised regression an dclassification (e.g., random forests, gradient boosting) and unsupervised clustering (e.g., K-means).

-   `TensorFlow`, `PyTorch` are **deep learning** libraries widely used for computer vision and natural language processing.

-   `optuna`, `Ray` can be integrated into ML/DL workflows for easy and efficient **model training, hyperparameter tuning, fine-tuning**, etc.

#### **Omics data analysis**

Emerging packages that provide standard omics data preprocessing and analysis pipelines allow Python to become increasingly popular in the field of bioinformatics:

-   `scanpy`, `anndata` are libraries for single-cell RNA-seq data loading, preprocessing, and analysis.

-   `Biopython` is a set of tools for biological computation that performs file parsering (BLAST, FASTA, GenBank, etc.), sequence analysis, clustering algorithms, etc.

-   `pysam` works with BAM/SAM/VCF files.

## Python vs R: Differences

While both programming languages are popular for data analysis and computation, Python and R differ in their underlying code structure, the scope of functionality, and the extensibility of tasks they can perform.
Here is a non-exhaustive summary of some key differences:

| Feature / Task | **R** | **Python** |
|-------------------|---------------------------|---------------------------|
| **Programming logic** | Mostly function-oriented ‚Äì structured around **functions** | Function- **AND** object-oriented ‚Äì structured around **classes** |
| **General-purpose programming** | ‚ö†Ô∏è Less ideal ‚Äì designed mainly for working with data | ‚úÖ Strong ‚Äì ML & AI, software development, scripting, etc. |
| **Computational power** | ‚úÖ Vectorization allows operating on all elements of a vector at once <br> ‚úÖ Best for statistical analysis <br> ‚ö†Ô∏è Memory-intensive; often slow for reading large data and performing large computations | ‚úÖ Generally faster for loops <br> ‚úÖ Strong support for GPU computing <br> ‚úÖ Memory-efficient for handling large objects and complex computations |
| **Package availability** | ‚úÖ Excellent for statistical analysis (`glm`, `survival`, `ggplot2`) <br> ‚ö†Ô∏è Good options for ML (`caret`, `mlr3`) but few DL packages <br> ‚úÖ Great for omics-focused analysis (`Bioconductor`, `ComplexHeatmap`, `Seurat`) | ‚òëÔ∏è Improving on statistical packages (`statsmodels`, `lifelines`) <br> ‚úÖ Best for ML/DL (`scikit-learn`, `pytorch`, `keras`) <br> ‚òëÔ∏è Fewer established packages specialized in omics data <br> ‚úÖ Great for processing large omics datasets (`scanpy`, `scvi-tools`) |
| **IDE & Reproducible environments (notebooks etc.)** | ‚úÖ RStudio <br> ‚úÖ RMarkdown, Quarto | ‚úÖ VS Code, JupyterLab, PyCharm, Spyder, etc. <br> ‚úÖ Jupyter Notebooks, Quarto |

## Integration

Python integrates with tools like:

-   **Jupyter Notebooks**: ideal for literate programming, sharing code + plots + narrative

-   **Snakemake / Nextflow**: workflow managers for scalable, reproducible bioinformatics pipelines

-   **Jupyer Notebook / Quarto**: documents that combine code, output, and markdown

-   **Git**: version control for collaboration

Next, we will

# 2 Installation and Setup

There are many ways to do install Python locally.
For the purpose of the workshop, we recommend one way of installation via [Miniconda](https://www.anaconda.com/docs/getting-started/miniconda/main) that works universally across Windows and MacOS platforms.
Miniconda is a minimal version of the [Anaconda Distribution](https://www.anaconda.com/docs/getting-started/anaconda/main), an open-source distribution of Python designed for scientific computing, data science, machine learning, and AI development.
It comes with **Python**, the **conda** package manager, and a minimal number of libraries.

Miniconda is relatively lightweight compared to Anaconda, which adds, on top of the above, the **Anaconda Navigator** **GUI** and over 300 pre-downloaded libraries.
While Anaconda can be a good alternative, we choose Miniconda for better efficiency in the installation process.

![Difference between conda, miniconda, and Anaconda](images/anaconda-explained.png){fig-align="left" width="400"}

| Feature | Anaconda | Miniconda |
|------------------|-------------------------------|------------------------|
| Size | ‚ö†Ô∏è\~3-4 GB (slow to download) | ‚úÖ\~400 MB (lightweight) |
| What is included? | conda, Python (latest version), 300+ popular packages, Anaconda Navigator | conda, Python (latest version), essential packages only |
| User-friendly? | ‚úÖGUI available (Anaconda Navigator) | ‚ö†Ô∏ècommand-line only ‚Äî `conda` |

: Anaconda vs. Miniconda

::: callout-tip
## Some ways to install Python

1.  Installing through the Python website
:::

## Installing Miniconda (Python+conda)

In this tutorial, we are going to install the **Minconda** and learn to use the features of its Anaconda Navigator desktop application.
This allows us to manage packages and environments without necessarily needing to know the conda terminal commands.

#### Step 1: Download the Minconda Installer

-   Download the installer from the Anaconda website: [https://www.anaconda.com/download/](https://www.anaconda.com/download/success#miniconda) (\*Make sure you are downloading under the "**Miniconda Installers**" section*)*

-   Choose the installer based on your system:

    -   **Windows**: `.exe` (64-bit or 32-bit)

    -   **macOS**: `.pkg` for GUI or `.sh` for terminal install

-   Or use the command line interface as preferred:

    ``` bash
    curl https://repo.anaconda.com/miniconda/Miniconda3-latest-Windows-x86_64.exe --output .\Downloads\Miniconda3-latest-Windows-x86_64.exe
    ```

#### Step 2: Install Miniconda

-   Go to your Downloads folder (or Home folder if downloaded via CLI) and double-click the installer to launch.
-   **On Windows:**

1.  Run the `.exe` installer

2.  Use default settings, but **check the box** that says:

    ‚úÖ *‚ÄúAdd Miniconda to my PATH environment variable‚Äù*

-   **On macOS:**

## Installing packages with Conda

Conda is a cross-platform package and environment manager that comes with Miniconda.
You can use conda to install packages from the Anaconda repositories (also called ***channels***).
It supports more than just Python packages but numerous other programming languages like R, Java, etc.

``` bash
conda install
```

::: callout-note
## Difference Between Conda and Pip

Long story short: ***Pip is for Python libraries only, while conda can install packages for any software (including python).***

It is generally recommended that you only use `conda install` when in a conda environment, as anything installed via pip won't be recognized by conda and vice versa.
Using the two interchangeably might overwrite or break packages and mess up the environment.

**What if the Python package is unavailable through conda?**

In these scenarios where the package you need is not built under conda, it makes sense to use pip to install packages within the conda environment.
The best practice is to ***install everything with conda first, then use pip if needed**.*

Check out this [blog](https://www.anaconda.com/blog/using-pip-in-a-conda-environment) for more information on using pip in a conda environment.
:::

You can also use Conda to create and manage virtual environments, install, or delete virtual environments.

## Virtual Environments

Environments are **isolated, independent installations** of a programming language and groups of packages that don't interfere with each other.

For example, you may have a Python version 3.8 installed on your computer as the System Python.
Meanwhile, you can install as many virtual environments as you want with the same or different Python versions and set of packages.

You can switch between environments for different projects, create environment files and share them with others.

## Why virtual environments?

You may find the flexibility of environments useful in many cases.

-   **Avoid Conflicts**.
    When you need libraries that are not compatibles with your system settings, such as an older Python version or conflicting dependencies.
    Creating a virtual environment can resolve the conflicts and changing it won't affect your other environments.

-   **Sharing Environment Setting**.
    You can also share your environment and the list of dependencies with someone with a copy of the `environment.yaml` file.

-   **Easy Management**.
    When your work is temporary or that you simply want to experiment things without having to worry about things breaking, you can work within a virtual environemnt and later delete it when needed.

## Creating a Virutal Environment with Conda

Alternatively, you can use the `conda` command line interface (CLI) to create virtual environments.

::: callout-tip
For users comfortable with command line, this approach is generally more recommended than Anaconda Navigator, as it is faster, more robust to broken environments (reported by some who used the GUI), and offers greater flexibility and functionality for environment management.
:::

1.  Create the virtual environment.

    ```         
    conda create --name <env-name>
    ```

    Replace `<env-name>` with the name you want to give your environment.

    *Note: you can use `-n` (shorthand) and `--name` interchangeably.*

2.  You can also specify the Python version and packages you want to install to your environment.

    ```         
    conda create -n <env-name> python=3.10 scipy=0.17.3 pandas matplotlib
    ```

    Or:

    ```         
    conda create -n <env-name> python=3.10
    conda install -n <env-name> scipy pandas matplotlib
    ```

    You can also specify channel other than the `defaults` channel (for multiple channels, pass the argument multiple times):

    ```         
    conda install -n <env-name> scipy --channel conda-forge --channel bioconda
    ```

3.  Now, activate your environment.

    ```         
    conda activate <env-name>
    ```

4.  You can also verify that your installation was successful by looking up the list of all current environments on your computer.

    ```         
    conda env list
    ```

    The default location for the installed conda environments (except for the base conda environment) is `..\anaconda3\envs\<env-name>`

5.  Deactivate the conda environment.

    Simply use `conda deactivate`.

    ::: callout-warning
    ## Note: Avoid activating on top of another virtual environment!

    Be careful when activating environments.
    Remember to always **deactivate the current environment before going into another one** because environments can be **stacked**.
    This can lead to chaos in the packages in both environments.
    So make sure that you see `(base)` at the beginning of the terminal prompt line when you are about to activate an environment.
    :::

6.  Removing an environment.

    -   Remove by environment name:

        ```         
        conda env remove -n <env-name>
        ```

    -   Remove by environment folder path:

        ```         
        conda env remove --prefix </path/to/your/env>
        ```

## 1.3.3 Creating an Environment from an `environment.yml` File

We can also create a virtual environment from a `.yml` file.

``` bash
conda env create -f environment.yml
```

Similarly, after installation, activate the new environment:

``` bash
conda activate <env-name>
```

This way, we can easily recreate an environment that is shared by others or share our environment settings with others.

For example, a simple environment file that has information about the environment name, channels, and dependencies:

```         
name: python310
channels:
  - defaults
dependencies:
  - python==3.10
  - pandas
  - numpy
```

Download the YML file for this Python workshop series [here](https://github.mskcc.org/Python-Workshop/Python-Workshop.github.io/tree/main/Downloadable/environment.yml).
This file includes the required channel and dependencies for completing the workshop sessions.

## Integrated Development Environment

and preferably set up a coding software for more elegant and interactive programming‚Äìwe often call it the integrated development environment (IDE).

::: callout-note
Rstudio is an IDE developed to support coding in both R and Python languages, with integrated functionalities including:

-   Automatic syntax highlighting, code completion, and smart indentation.

-   Multi-panel layout to display coding scripts, console, environment variables, and plot side-by-side.

-   Support for RMarkdown and Quarto documents.

-   Project management through Rprojects and renv.
:::

Ther are many IDEs compatible with Python programming **Visual Studio Code JupyterLab** PyCharm, Spyder

| Tool | Description |
|------------------|------------------------------------------------------|
| **VS Code** | Lightweight, powerful IDE (extensible with Python & Jupyter extensions) |
| **JupyterLab** | Interactive notebooks for analysis & reports |
| **Spyder** | RStudio-like interface, good for scientific Python |
| **PyCharm** | Full-featured Python IDE (more for software dev) |

## Visual Studio Code

![](images/vs-code-icon.png){width="100"}

> **Visual Studio Code (VS Code)** is one of the most popular open-source code editors with many features.

-   **Multi-Language Programming.** You can code in almost any major programming languages in VS Code including Python, R, C/C++, JavaScript, etc.
-   **Build-In Git Source Control.** VS Code automatically recognizes and uses the computer's Git installation to allow project version control. You can easily track changes, stage, and commit changes to your working branch.
-   **Variety of Project Development Support**. You can add extra features such as language packs, debugging tools, Git/Github features, and remote server connector by installing extensions from the Extension Marketplace.

Go to [1.4 Work with Jupyter/Quarto in VS Code] for guide on how to get started with VS Code.

## Quarto

![](images/quarto-icon.png){width="250"}

> **Quarto** is an open-source scientific and technical publishing system.

Think of Quarto `.qmd` as similar to R Markdown `.Rmd` files.
Both combine executable code chunks with text components and figures and allow generating outputs as PDF, HTML, Docx, and even slideshow presentations.

But Quarto has more:

-   **Compatibility with multiple IDEs**.
    You can not only work with Quarto `.qmd` in RStudio, but also VS Code, Jupyter Lab, etc.

-   **Mutli-lingual support**.
    Unlike R Markdown which is dependent on R, Quarto does not require R.
    It supports embedded Python, JavaScript, and Julia executable code by simply specifying the language name in the braces on top of a code chunk (e.g., ```` ```{python} ````).

-   **Multi-engine support**.
    Don't worry if you are an R Markdown or Jupyter Notebook user!
    Qurato also works with `.Rmd` and `.ipynb` files and will automatically deploy either the knitr or jupyter engine depending on the file type that you are rendering.

In summary, Quarto is easy for R users to transition into due to similarities with the R Markdown.
It also includes more functionality and flexibility, making it a great tool for learning Python.

::: callout-note
## Ecosystem of Python Programming

-   **Miniconda** provides a lightweight way to manage Python environments, similar to renv in R, ensuring dependency isolation without cluttering the global system.
    It simplifies package management, much like CRAN + Bioconductor, while also handling system-level dependencies.

-   **VS Code**, akin to RStudio, offers a powerful yet customizable IDE with Python-specific extensions, an interactive Jupyter notebook interface, and seamless Git integration.

-   Together, Miniconda and VS Code provide an efficient, modular setup, making Python adoption intuitive while maintaining the reproducibility and project structure familiar to R users.
:::

# 1.4 Work with Jupyter/Quarto in VS Code

## Prerequisites

#### Install Everything

1.  Install **VS Code**: [Download and install VS Code](https://code.visualstudio.com/download)
2.  Install **Quarto**: [Install from the official website](https://quarto.org/docs/get-started/)
3.  Install **Conda**: Go to previous section \[1.2 Install Anaconda\].

## Setting up VS Code

1.  Open VS Code.

2.  **Install Extensions from Extension Marketplace.** Click **Extensions** from the left toolbar or click `Ctrl+Shift+X` (or `Cmd+Shift+X` on Mac).

    -   **Python**: To support Python language, debugging, documentations, etc.

    -   **Quarto**: To support Quarto document editing and previewing.

    -   **Jupyter**: To support rendering Python documents from Jupyter Notebooks or Quarto files.

    ::: callout-important
    You still need to install conda (Python) and Quarto to your computer to enable these extensions.
    :::

3.  **Create a Conda Virtual Environment.** You may also use existing ones you created.

    -   Required packages/dependencies (to install, use `conda install <package-name> ‚Äìchannel conda-forge`):

        -   jupyter
        -   ipykernel
        -   pyyaml

    <!-- -->

    -   **Note**: we recommend installing an environment from this [YML file](https://github.mskcc.org/Python-Workshop/Python-Workshop.github.io/tree/main/Downloadable/environment.yml). See \[1.3.3 Creating an Environment from an environment.yml File\] for detailed instructions.

4.  **Configure the Environment in VS Code.**

    1.  Open Command Palette by pressing `Ctrl+Shift+P` (or `Cmd+Shift+X` on Mac).

    2.  Type "**Python: Select Interpreter**".

    3.  Choose the Conda python environment you created.
        If you don't see its name pop up, choose **Enter interpreter path...** and manually type in the location of your conda virtual environment python executable.

        E.g., `C:\Users\<username>\AppData\Local\anaconda3\envs\<env-name>\python.exe` for Windows **or** `Users/<username>/anaconda/envs/<env-name>/python` for macOS.

        ::: callout-note
        ## Find your conda Python executable path

        You can search for the conda Python interpreter location on your computer.
        Here is an example of how to do it using command line.

        For both Windows and macOS, open the **Anaconda Prompt** or **terminal** and activate the conda Python environment with `conda activate <env-name>`.

        Then, locate your Python executable by typing the following:

        -   Windows: `where python`

        -   macOS: `which python`
        :::

Now you are ready to go.
Let's start with creating a new Quarto file.

## Create a Quarto File

Go to **`File > New File`** and select Quarto Document.
This will generate an empty `.qmd` file with the following YAML metadata.
You can change the title and file output format as needed.

``` markdown
---
title: "First Day"
format: html
---
```

In Quarto files, you can mix Markdown with code blocks just like in R Markdown.
For example:

``` markdown
# Heading 1
This is a demo Quarto file. Here is the text section.

## Heading 2
This is a subsection.
```

To add an executable code block, specify the language name inside two curly braces (e.g., `{python}`, `{r}`, etc.).
For example, let's print something:

```` markdown
```{{python}}
print("This is some Python output.")
```
````

```{python}
# | echo: false
print("This is some Python output.")
```

#### File Previewing

To preview the html output, click on the **Quarto Preview** button at the top right (or press `Ctrl+Shift+K`).

Alternatively, run `quarto preview` in your terminal to view the output.

``` bash
quarto preview <yourfilename>.qmd
```

This is an example of the html output preview showing next to the source `.qmd` file.

![](images/quarto-demo.png)

#### File Rendering/ Exporting to other Formats

To save the output, use **Quarto: Render Document** from the command search bar (`Ctrl+Shift+P`).

![](images/quarto-render.png){fig-align="left" width="900"}

Then select the desired output format.
For example, Render HTML.

![](images/quarto-render2.png){width="900"}

Similarly, you can use command line `quarto render` in the terminal to save the output.

``` bash
quarto render <yourfilename>.qmd
```

Now you should see a `<yourfilename>.html` file generated in your working directory.